### ISSUE TITLE
[Bug] Shape mismatch in RT-DETR/D-FINE loss due to premature initialization in BaseTrainer.setup()

### ISSUE THREAD

**Issue by bencejdanko** [Libre-YOLO/libreyolo#319]
The initialization order in `BaseTrainer.setup()` is:

1. `self.on_setup()` (where subclass trainers construct their loss module/criterion).
2. `self._setup_data()` (loads the dataset's `data.yaml` and is the first time the real dataset class count `nc` is known). If the dataset's `nc` is different from the wrapper model's default, it rebuilds the model head to match and updates `self.config.num_classes`.

* **YOLO-grid models (YOLO9, YOLOX):**  Because their loss calculation is set inside the model modules (e.g., `yolo9/model.py` and `yolox/model.py`). Rebuilding the head automatically replaces the loss structures with the active head channels

* **RT-DETR / D-FINE / DEIM:** These families have a bug: 
  * In `rtdetr/trainer.py`, the `on_setup` method initializes the DETR loss module:
    ```python
    def on_setup(self):
        self.criterion = RTDETRLoss(num_classes=self.config.num_classes)
        self.criterion.to(self.device)
    ```
  * In `rtdetr/loss.py`, the criterion registers target label buffers based on the initial class count:
    ```python
    target = F.one_hot(target_classes, num_classes=self.num_classes + 1)[..., :-1]
    ```

The risk is if a user trains RT-DETR with a custom dataset having an `nc` different from the default model config (e.g., custom 3-class dataset vs. default 80 COCO classes), `RTDETRLoss` is created with `num_classes=80`, but the model head is rebuilt for `3` classes. On the first training batch the loss forward pass will feed a 3-channel predicted tensor into a one-hot encoder expecting 80 channels, crashing with a shape mismatch error

**Comment by EHxuban11** [comment]
Hello @bencejdanko , thank you for the issue!!

I can't reproduce the crash through the API because `model.train(data=...)` reads the dataset yaml one level earlier, before the trainer is constructed. So normal users don't have to worry about this.

But yeah, I agree that the loss code is fragile for developers.

**Comment by EHxuban11** [comment]
I'm working now on hardening it in this branch:
https://github.com/LibreYOLO/libreyolo/tree/319-bug-shape-mismatch-in-rt-detrd-fine-loss-due-to-premature-initialization-in-basetrainersetup
